src/libs/elektra/kdb.c

```diff
@@ -2465,6 +2472,17 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
 	// Step 12c: run postcommit phase
 	runSetPhase (backends, parentKey, KDB_SET_PHASE_POST_COMMIT, true, KDB_SET_FN_COMMIT);
 
+
+
+	struct _SendNotificationHook * sendNotificationHook = handle->hooks.sendNotification;
+	while(sendNotificationHook != NULL)
+	{
+		// TODO (atmaxinger): Is setKs really the correct KeySet?
+		// Unfortunately setKs removes the SYNC flag, so plugins can not reliably detect changed keys ...
```

Comment:
This is something we need to tackle. As of now the plugins are called, but for example dbus does not correctly detect changed keys, as it relies on the `keeNeedsSync` function.

Maybe we change the contract and provide the plugins with the three KeySets (added, changed, removed) directly?

Reply:
> I thought about this too, but we'd also need to map it using the parent keys used to get the keysets. Otherwise we would only have the last retrieved keyset stored, and there is no way of enforcing that the user of the API will only perform `kdbGet` after a `kdbSet` of the previous read keyset.

I'm not entirely sure what you mean here. But regardless wouldn't any such issue exist in exactly the same way in a plugin?

It is already mandatory to call `kdbGet` before `kdbSet` and the `kdbGet` call must use the same or more backends as `kdbSet` needs.

There is also the `KeySet * keys` in each of the `struct _BackendData` in `kdb->backends`. I'd have to check, but maybe these already are or at least could be `keyDup`'ed and maybe they aren't cleared at the end of `kdbGet`. Then we could use that data at the start of `kdbSet` (before all the `KeySet * keys` are overwritten).


---

> This mostly works, but now all keys in `setKs` are marked as `SYNC`. Probably because the keys are marked somewhere in the process of `backendsDivide`. I think this is because of the `keyDup` calls here:

Yeah, I forgot about that. `ksDeepDup` actually restores the sync flag after calling `keyDup`.

> Maybe we should set the SYNC flag only when it is there on the original key?

So yes, we have to do something similar. Instead of just `keyDup (k, KEY_CP_ALL)` we need:

```c
Key * d = keyDup (k, KEY_CP_ALL);
clear_bit (d->flags, KEY_FLAG_SYNC);
d->flags |= k->flags & KEY_FLAG_SYNC;
```

Reply:
Yes, the same problem persists in the plugin. What I mean is this:

```
Key * keyA = ...;
Key * keyB = ...;

KeySet * a = ...;
KeySet * b = ...;

kdbGet(kdb, a, keyA);
kdbGet(kdb, b, keyB);

// Change keys in a

kdbSet(kdb, a, keyA);
```
If we only store the last KeySet that has been retrieved, we would have stored `b`, but now we are updatding KeySet `a`.

Reply:
I created issue #4514 for this discussion.